R4R: prometheus exporter in substrate (#4511)
paritytech/substrate: utils/prometheus/README.md

> # Substrate Prometheus Node Exporter

Author: complete

> ![grants](./photo_2019-12-13_16-32-53.jpg)

Author: This is a request from the W3F, which can only be placed in a private repo under consultation

## Introduction

Prometheus is one of the most widely used monitoring tool for managing high availability services supported by [Cloud Native Computing Foundation](https://www.cncf.io/). By providing Prometheus metrics in Substrate, node operators can easily adopt widely used display/alert tools such as Grafana and Alertmanager without setting-up/operating external Prometheus push gateways (which is an antipattern in the first place) through RPC connections. Easy access to such monitoring tools will benefit parachain developers/operators and validators to have much higher availability of their services.

## Table of Contents

Hack Prometheus in Substrate
 - Prometheus primer
 - CLI Config
 - Metrics Add

Metrics
 - List of available metrics

Start Prometheus
 - Install prometheus
 - Edit Prometheus config file
 - Start Prometheus

Start Grafana
 - Install Grafana

## Metrics

substrate can report and serve the Prometheus metrics, which in their turn can be consumed by Prometheus collector(s).

This functionality is disabled by default.

To enable the Prometheus metrics, set in your cli command (--prometheus-addr,--prometheus-port ). 

> Metrics will be served under /metrics on 33333 port by default.

Reviewer: Yup

> Metrics will be served under /metrics on 33333 port by default.

Author: complete

> ### List of available metrics

Author: complete

Consensus metrics, namespace: `substrate`

| **Name**                               | **Type**  | **Tags** | **Description**                                                 |
| -------------------------------------- | --------- | -------- | --------------------------------------------------------------- |

> | consensus_finality_block_height_number | IntGauge  |          | finality Height of the chain                                    |

Author: complete

| Resource_cpu_use(Future)               | IntGauge  |          | Operating System cpu load                                       |
| Resource_disk_use(Future)              | IntGauge  |          | Operating System disk use                                       |
| validator_sign_prevote(Future)         | IntGauge  | validator addr | validator sign vote list                               	  |
| validator_sign_precommit(Future)       | IntGauge  | validator addr | validator sign commit list                                |

## Start Prometheus
### Install prometheus

https://prometheus.io/download/
```bash
wget <download URL>
tar -zxvf <prometheus tar file>
```

### Edit Prometheus config file

You can visit [prometheus.yml](https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus.yml) to download default `prometheus.yml`.

Then edit `prometheus.yml` and add `jobs` :

```yaml
      - job_name: kusama
          static_configs:
          - targets: ['localhost:33333']
            labels:
              instance: local-validator
```

> Note：value of targets is ip:port which used by substrate monitor 

### Start Prometheus

```bash

> cd <prometheus file>

Author: complete

> # Substrate Prometheus Exporter

Reviewer 2: What do you think of the following changes?

```diff
diff --git a/utils/prometheus/README.md b/utils/prometheus/README.md
index c6746d573..d7c895c3f 100644
--- a/utils/prometheus/README.md
+++ b/utils/prometheus/README.md
@@ -2,44 +2,25 @@
 
 ## Introduction
 
-Prometheus is one of the most widely used monitoring tool for managing highly available services supported by [Cloud Native Computing Foundation](https://www.cncf.io/). By providing Prometheus metrics in Substrate, node operators can easily adopt widely used display/alert tools such as Grafana and Alertmanager without setting-up/operating external Prometheus push gateways (which is an antipattern in the first place) through RPC connections. Easy access to such monitoring tools will benefit parachain developers/operators and validators to have much higher availability of their services.
+[Prometheus](https://prometheus.io/) is one of the most widely used monitoring
+tools for managing highly available services supported by [Cloud Native
+Computing Foundation](https://www.cncf.io/). By providing Prometheus metrics in
+Substrate, node operators can easily adopt widely used display/alert tools such
+as [Grafana](https://grafana.com/) and
+[Alertmanager](https://prometheus.io/docs/alerting/alertmanager/). Easy access
+to such monitoring tools will benefit parachain developers/operators and
+validators to have much higher availability of their services.
 
-## Table of Contents
+Metrics will be served under `/metrics` on TCP port 9615 by default.
 
-Metrics
 
-Start Prometheus
- - Edit Prometheus config file
+## Quick Start
 
+1. From the root of the repository start Substrate `cargo run --release`.
 
-## Metrics
+2. In another terminal run `curl localhost:9615/metrics` to retrieve the
+   metrics.
 
-Substrate can report and serve the Prometheus metrics, which in their turn can be consumed by Prometheus collector(s).
 
-This functionality is disabled by default.
-
-Metrics will be served under /metrics on 9615 port by default.
-
-
-## Start Prometheus
-### Edit Prometheus config file
-
-You can visit [prometheus.yml](https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus.yml) to download default `prometheus.yml`.
-
-Then edit `prometheus.yml` and add `jobs` :
-
-```yaml
-      - job_name: kusama
-          static_configs:
-          - targets: ['localhost:9615']
-            labels:
-              instance: local-validator
-```
-
-> Note：value of targets is ip:port which used by substrate monitor 
-
-### Start Prometheus
-
-> The above example, you can save `prometheus.yml` at `~/volumes/prometheus` on your host machine
-
-You can visit `http://localhost:9090` to see prometheus data.
+To learn how to configure Prometheus see the Prometheus [Getting
+Started](https://prometheus.io/docs/prometheus/latest/getting_started/) guide.
```

Author: OK, simple is the best

# Substrate Prometheus Node Exporter
![grants](./photo_2019-12-13_16-32-53.jpg)
## Introduction

> Prometheus is one of the most widely used monitoring tool for managing high availability services supported by [Cloud Native Computing Foundation](https://www.cncf.io/). By providing Prometheus exporter in substrate, node operators can easily adopt widely used display/alert tool such as Grafana without seting-up/operating external Prometheus agent through RPC connections. Easy access to such monitoring tools will benefit parachain develepers/operators and validators to have much higher availability quality of their services.

Reviewer 2: - Nit: Given that this is within the binary, one would not speak of a Prometheus exporter, but a Prometheus metrics endpoint.

- Replacing `Prometheus agent` with `Prometheus push gateway`.

- Various typos.

```suggestion
Prometheus is one of the most widely used monitoring tool for managing high availability services supported by [Cloud Native Computing Foundation](https://www.cncf.io/). By providing Prometheus metrics in Substrate, node operators can easily adopt widely used display/alert tools such as Grafana and Alertmanager without setting-up/operating external Prometheus push gateways (which is an antipattern in the first place) through RPC connections. Easy access to such monitoring tools will benefit parachain developers/operators and validators to have much higher availability of their services.
```

Author: complete

> ## List of Contents

Reviewer 2: ```suggestion
## Table of Contents
```

Author: complete

Hack Prometheus in Substrate

>  - Prometheus starter

Reviewer 2: What is a `Prometheus starter`? Maybe a `Prometheus 101` or a `Prometheus primer`?

Author: complete

 - CLI Config
 - Metrics Add

Metrics
 - List of available metrics

Start Prometheus
 - Install prometheus
 - Edit Prometheus config file
 - Start Prometheus

Start Grafana
 - Install Grafana

## Substrate Dev hack
### Prometheus starter

Here is the entry point of prometheus core module in Parity Substrate.

> In existing sources, refer to the grapana source due to the issue of the wasm.

Reviewer 2: ```suggestion
In existing sources, refer to the Grafana source due to the issue of the wasm.
```

Author: complete

> In existing sources, refer to the grapana source due to the issue of the wasm.

Reviewer 2: Can you add some more details here, I am not sure I follow the intention of the comment.

utils/prometheus/src/lib.rs
```rust

> #[macro_use]

Reviewer 2: I am not sure I understand the reasoning for putting all of this source code here. It is very large as an example and difficult to maintain as there is no static analysis of code snippets within Markdown as far as I am aware.

				},
        _ => return Err(err.into())
      }
		}
	};
  let service = make_service_fn(|_| {
		async {
			Ok::<_, Error>(service_fn(request_metrics))
		}
	});

	let _server = Server::builder(Incoming(listener.incoming()))
		.executor(Executor)
		.serve(service)
    .boxed();
  
  
	let result = _server.await.map_err(Into::into);

	result
}

#[cfg(target_os = "unknown")]
pub async fn init_prometheus(_: SocketAddr) -> Result<(), Error> {
	Ok(())
}

#[macro_export]
macro_rules! prometheus_gauge(
  ($($metric:expr => $value:expr),*) => {
    use $crate::{metrics::*};
    $(
        metrics::set_gauge(&$metric, $value);
    )*
  }
);

#[macro_export]
macro_rules! prometheus_histogram(
  ($($metric:expr => $value:expr),*) => {
    use $crate::{metrics::*};
    $(
        metrics::set_histogram(&$metric, $value);
    )*
  }
);

/*
TODO: Make abstract type for all metrics(e.g. Gauge, Histogram, Counter) with generic traits so that all metrics can be set up with one function `set`
#[macro_export]
macro_rules! prometheus(
  ($($a: expr; $metric:expr => $value:expr),*) => {
    use $crate::{metrics::*};
    $(
        metrics::set(#$a, &$metric, $value);
    )*
  }
);
*/
```

> Here is the dependancies of the module.	

Reviewer 2: ```suggestion
Here are the dependencies of the module.	
```

| p2p_peers_number                       | IntGauge  |          | Number of peers node's connected to                             |
| p2p_peer_receive_bytes_per_sec         | IntGauge  |          | number of bytes received from a given peer                      |
| p2p_peer_send_bytes_per_sec            | IntGauge  |          | number of bytes sent to a given peer                            |
| Resource_receive_bytes_per_sec(Future) | IntGauge  |          | Operating System of bytes received                              |
| Resource_send_bytes_per_sec(Future)    | IntGauge  |          | Operating System of bytes sent                                  |
| Resource_cpu_use(Future)               | IntGauge  |          | Operating System cpu load                                       |
| Resource_disk_use(Future)              | IntGauge  |          | Operating System disk use                                       |
| validator_sign_prevote(Future)         | IntGauge  | validator addr | validator sign vote list                               	  |
| validator_sign_precommit(Future)       | IntGauge  | validator addr | validator sign commit list                                |

> ## Start Prometheus

Reviewer 2: I don't think this is worth duplicating here. I would suggest to link to the Prometheus docs.

Author: I agree. I'm dropping the table

Author: complete

> ### Install prometheus

Reviewer 2: ```suggestion
### Install Prometheus
```

Author: complete

https://prometheus.io/download/
```bash
wget <download URL>
tar -zxvf <prometheus tar file>
```

### Edit Prometheus config file

You can visit [prometheus.yml](https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus.yml) to download default `prometheus.yml`.

Then edit `prometheus.yml` and add `jobs` :

```yaml
      - job_name: kusama
          static_configs:
          - targets: ['localhost:33333']
            labels:
              instance: local-validator
```

> Note：value of targets is ip:port which used by substrate monitor 

### Start Prometheus

```bash
cd <prometheus file>
./prometheus
```

> The above example, you can save `prometheus.yml` at `~/volumes/prometheus` on your host machine

You can visit `http://localhost:9090` to see prometheus data.

## Start Grafana
### Install Grafana

> https://grafana.com/docs/installation/debian/

Reviewer 2: Again, I don't think it is worth explaining how to install Grafana nor Prometheus.

Reviewer: Agreed

Author: ok, Agreed !Milestone 2 will provide a fully configured dashboard, so we will consider removing it in this release and tagging grafana official links the next release.

```bash
apt-get install -y software-properties-common
sudo add-apt-repository "deb https://packages.grafana.com/oss/deb stable main"
wget -q -O - https://packages.grafana.com/gpg.key | sudo apt-key add -
sudo apt-get update
sudo apt-get install grafana
sudo service grafana-server start
./prometheus
```

You can visit `http://localhost:3000/` to open grafana and create your own dashboard.

> Tips: The default username and password are both admin. We strongly recommend immediately changing your username & password after login

> ### Seting Grafana

Reviewer 2: ```suggestion
### Setting up Grafana
```

Author: complete
